Python - Efficient concatenation with `join()` (#1588)
enkidevs/curriculum: python/core/working-with-strings/efficient-concatenation-with-join.md

---

# Efficient concatenation with `join()`

---
## Content

> With strings being immutable in Python, there are certain advantages or disadvantages present.

Reviewer: Maybe:

```suggestion
With strings being immutable in Python, there are certain advantages and disadvantages to consider.
```

> An advantage is represented by the fact that strings can be used directly as keys in dictionaries. On top of that, string copies can be shared amon multiple variable bindings.

Reviewer: ```suggestion
An advantage is represented by the fact that strings can be used directly as keys in dictionaries. On top of that, string copies can be shared among multiple variable bindings.
```

> One disadvantage is that if you want to amend something in an existing string, then you have to create a new one. This leads to significant inefficiencies.

Reviewer: ```suggestion
One disadvantage is that if you want to amend any part of an existing string, you must create a new one with that change. This means that programs can easily lead to creation of many unnecessary strings, which is quite inefficient and can significantly increase the memory usage.
```

> In this insight we will present you several examples of such inefficiencies and how you can overcome them.

Reviewer: Seems like there's only one example bellow?

Maybe we should say:

```suggestion
Let's see a few examples of how to update strings more efficiently and prevent unnecessary memory usage.
```

> Let's say you would like to build a string sequentially:

Reviewer: ```suggestion
For instance, a naive approach to combining a list of strings into a single string would be to use a loop and string concatenation:
```

Can also maybe add a footnote at "string concatenation"

* Enki
* E n k i
* E,n,k,i

---
## Footnotes

[1:Efficient Concatenation]
To understand why iteratively creating a string with `+=` is inefficient, let's take a look at this example:
```py
x = 'en'
x += 'ki' # 'enki' 
x += 'enki' # 'enkienki'
```

> In this example, Python first allocates and creates `'enki'`, before it can allocate and create `'enkienki'`. When concatenating a small number of strings this wouldn't pose a problem, but what would happen if the total number of string that need to be concatenated was over 1000? In this case, each substring would get copied approximately `N/2` times (where `N` represents the number of strings to be concatenated). You can already see why this method does not scale well with a large number of strings.

Reviewer: How about?

```suggestion
In this example, Python first allocates the memory for `'en'`, then again for `en` and `ki` (to make `enki`), then again for `en` and `ki` and `enki` (to make `'enkienki'`). This means that each string from the previous concatenation recreates more memory for the next concatenation even though it already had enough memory to fit in. In this scenario, instead of just creating the memory for `enkienki`, we would end up creating that memory plus the memory needed for `en` and `enki` during prior concatenations. 
When concatenating a small number of strings this wouldn't pose much of a problem, but what would happen if the total number of strings that need to be concatenated was 1000? Since each substring gets re-created for all concatenations that follow it, the first substring would get created 1000 times, the second one 999 times, the third one 998 times, and so forth, which averages to about ~500 concatenations per substring. This means that concatenation of 1000 strings  will end up with approximately 500,000 concatenations! In general terms, concatenating `N` strings would lead approximately to `(N*N)/2` concatenations, which is extremely inefficient. This is why we must handle Python's string immutability with caution.
```

> On the other hand, the `join()` method uses some tricks to find out the memory that needs to be allocated and copying each substring to the new buffer. This means that each string is copied only once, resulting in a performance increase.

Reviewer: ```suggestion
On the other hand, the `join()` method uses some internal tricks to find out the memory that needs to be allocated ahead of time and copies each substring to the new buffer. This means that each string is copied only once, making the operation ideally efficient. Concatenating `N` strings using `join()` leads to `N` concatenations, which is what one would expect.
```
